ethcore/private-tx/src/state_store.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::{Instant, Duration};
+use parking_lot::RwLock;
+use ethereum_types::H256;
+use kvdb::KeyValueDB;
+use types::transaction::SignedTransaction;
+use private_transactions::VerifiedPrivateTransaction;
+use private_state_db::PrivateStateDB;
+use log::Logging;
+
+/// Max duration of retrieving state (in ms)
+const MAX_REQUEST_SESSION_DURATION: u64 = 120 * 1000;
+
+struct HashRequestSession {
+	hash: H256,
+	expiration_time: Instant,
+}
+
+/// Type of the stored reques
+pub enum RequestType {
+	/// Verification of private transaction
+	Verification(Arc<VerifiedPrivateTransaction>),
+	/// Creation of the private transaction
+	Creation(SignedTransaction),
+}
+
+#[derive(Clone, PartialEq)]
+enum RequestState {
+	Syncing,
+	Ready,
+}
+
+struct StateRequest {
+	request_type: RequestType,
+	request_hashes: HashSet<H256>,
+	state: RequestState,
+}
+
+/// Wrapper over storage for the private states
+pub struct PrivateStateStorage {
+	private_state_db: Arc<PrivateStateDB>,
+	requests: RwLock<Vec<StateRequest>>,
+	syncing_hashes: RwLock<Vec<HashRequestSession>>,
```

Comment:
it seems like you mostly need find by `hash` for `syncing_hashes`, 
so why not just use `RwLock<HashMap<H256, Instant>>`? (Or `H256FastMap`)
It would make implementation simpler and more readable and decrease asymptotic complexity.

Reply:
Reworked